### PR TITLE
Publish the build independent from the integration test results

### LINKED ISSUE #2329: serviceaccount "rook-ceph-cluster" not found
`kubectl create -f cluster.yaml`
 
cloud not create `osd` pods

` kubectl describe jobs.batch -n rook-ceph rook-ceph-osd-prepare-xxx`
```
Name:           rook-ceph-osd-prepare-vm-16-6-ubuntu
Namespace:      rook-ceph
Selector:       controller-uid=5c3eca4d-f7ac-11e8-98a6-525400842c0a
Labels:         app=rook-ceph-osd-prepare
                rook_cluster=rook-ceph
Annotations:    <none>
Parallelism:    1
Completions:    1
Pods Statuses:  0 Running / 0 Succeeded / 0 Failed
Pod Template:
  Labels:           app=rook-ceph-osd-prepare
                    controller-uid=5c3eca4d-f7ac-11e8-98a6-525400842c0a
                    job-name=rook-ceph-osd-prepare-vm-16-6-ubuntu
                    rook_cluster=rook-ceph
  Service Account:  rook-ceph-cluster
  Containers:
   copy-bins:
    Image:      rook/ceph:master
    Port:       <none>
    Host Port:  <none>
    Args:
      ceph
      osd
      copybins
    Environment:
      ROOK_PATH:  /rook
    Mounts:
      /rook from rook-binaries (rw)
   provision:
    Image:      ceph/ceph:v13
    Port:       <none>
    Host Port:  <none>
    Command:
      /rook/tini
    Args:
      --
      /rook/rook
      ceph
      osd
      provision
    Environment:
      ROOK_NODE_NAME:             vm-16-6-ubuntu
      ROOK_CLUSTER_ID:            3f729e07-f7ac-11e8-98a6-525400842c0a
      ROOK_PRIVATE_IP:             (v1:status.podIP)
      ROOK_PUBLIC_IP:              (v1:status.podIP)
      ROOK_CLUSTER_NAME:          rook-ceph
      ROOK_MON_ENDPOINTS:         <set to the key 'data' of config map 'rook-ceph-mon-endpoints'>  Optional: false
      ROOK_MON_SECRET:            <set to the key 'mon-secret' in secret 'rook-ceph-mon'>          Optional: false
      ROOK_ADMIN_SECRET:          <set to the key 'admin-secret' in secret 'rook-ceph-mon'>        Optional: false
      ROOK_CONFIG_DIR:            /var/lib/rook
      ROOK_CEPH_CONFIG_OVERRIDE:  /etc/rook/config/override.conf
      ROOK_OSD_DATABASE_SIZE:     1024
      ROOK_OSD_JOURNAL_SIZE:      1024
      ROOK_DATA_DIRECTORIES:      /rook/storage-dir
    Mounts:
      /etc/ceph from ceph-default-config-dir (rw)
      /rook from rook-binaries (rw)
      /rook/storage-dir from rook-storage-dir (rw)
      /var/lib/rook from rook-data (rw)
  Volumes:
   rook-data:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/rook
    HostPathType:
   ceph-default-config-dir:
    Type:    EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
   rook-config-override:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      rook-config-override
    Optional:  false
   rook-binaries:
    Type:    EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
   rook-storage-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /rook/storage-dir
    HostPathType:
Events:
  Type     Reason        Age   From            Message
  ----     ------        ----  ----            -------
  Warning  FailedCreate  9s    job-controller  Error creating: pods "rook-ceph-osd-prepare-vm-16-6-ubuntu-" is forbidden: error looking up service account rook-ceph/rook-ceph-cluster: serviceaccount "rook-ceph-cluster" not found
```

### PR DESCRIPTION
**Pull request by travisn**
Signed-off-by: travisn <tnielsen@redhat.com>

**Description of your changes:**
This change will cause master and release builds to be published before the integration tests are run. There are some positive and negative aspects to this:

PROS:
- Published builds match what has been merged
- Random integration test failures don't block the publishing of the build

CONS:
- If there is a significant failure shown by the integration tests, the build will be published anyway

I wouldn't be surprised if there is much debate around this, but the reality of the situation is that the random failures in integration tests are too painful not to do this. It doesn't mean we deprioritize fixing the integration test issues, but this will help us ensure our published images match what has been merged to the repo. I'm not worried about the integration tests finding significant issues after the merge, because they would have already been vetted **before** the merge.

As an example, we haven't had a successful master build for 6 days, but there was a breaking change three days ago that requires a new build. See #2329.

**Which issue is resolved by this Pull Request:**
Resolves #2329 

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md#comments)

### CODE REVIEW

**Comment by travisn** [comment]
@jbw976 I also struggled with it at first glance. But the more I've thought about it, the more it seems right because what is merged will be immediately published. This is needed so the docs immediately match what is merged, as well as publishing a master image that matches the yamls in master. A couple times this week we've had users hit issues where there was a long delay between the updated yamls and the image getting published due to random build issues.

Since we've already validated before the merge, there is a very low chance something will be failing after the merge. I can't remember the last time we had a real issue in master or a release branch. 

The publish step is a no-op in PR branches. Neither the docs nor the image will be published. There are conditions on the actions inside the publish step that prevent them from executing.

**Comment by bassam** [comment]
Seems odd that we would publish artifacts that have not passed our tests. It feels like we're pushing the problem somewhere else.

Is there a way to flag certain tests as intermittent? We obviously need to take on this problem in full force for the next release.

**Comment by travisn** [comment]
@bassam Not sure how to mark a test as intermittent. But if we did, that would result in a green build and we would end up ignoring the failures. This way we can still see the failures. And if we decide there really is an issue with a release build, it should fail our manual release criteria and we would rev up a new build with a fix. We already do this manual validation anyway before release.

**Comment by bassam** [comment]
typically intermittent tests should still show as a failure or warnings, but do not stop the rest of the build.

**Comment by BlaineEXE** [comment]
This does make me a bit nervous, but I'm not totally opposed. There could potentially be an unexpected difference in behavior in some extreme corner case following a git merge, but that does seem unlikely. For publishing master builds, I think this change seems okay.

But I would want to set a process expectation in place that the tests all have to pass before releasing any versioned release (e.g., `v0.9.1`, `v0.8.4`).

**Comment by travisn** [comment]
I'd like to make another argument for merging this. We have had two major CI issues this week. Their impact would have been much lessened if this had been in place.
1. GCE load time blocked integration tests for over a day. Nothing could be merged to master. But if we published before integration tests in master, we could have merged to master and had a corresponding build published.
1. GCE instances not getting cleaned up. This affected both master on Monday and the release build today. We couldn't test the release build specifically because the build hadn't been published.  

And even if the CI were perfect, I would still argue this is the correct behavior. Whatever is merged to a branch should have a corresponding published image, even if there is some integration test failing. Whether or not to notify users that the build is ready for consumption is a separate question.

**Comment by BlaineEXE** [comment]
I'm inclined to agree with Travis' reasoning about each master commit having a corresponding image. I would propose to additionally have master builds changed so that they no longer build with the tag `v0.9.0-gHASH` but instead show up as `master-gHASH` for user clarity. This is very non-intuitive naming, and I was confused when I first encountered it.

**Comment by travisn** [comment]
@bassam @jbw976 Once again today we were bit by the master build not getting published and a user running into an issue because of that. It really is a recurring nightmare not to have builds published that match what has been merged. Can we merge this?

**Comment by bassam** [comment]
I think it would be ok to `publish` the build before integration testing, as long as we do not `promote` it.

**Comment by travisn** [comment]
@bassam `promote` is run as a separate pipeline for release builds, right? Is that sufficient or are you saying another change is needed for this PR?

**Comment by bassam** [comment]
> Before a PR is merged, it must pass integration tests in its branch so we would not expect any failures afterwards.
> After the merge, if there is any conflict with some other recent merge the likelihood of failing integration tests after passing build and unit tests is very low.

Yes but the PR does not have all the changes in master, right? We don't require all branches to be up to date before merging currently. So if two PRs were submitted and both passed in their respective branches and merged, the first time they are tested together is in master.

When the integration tests fail in master it's hard to tell if it's because of a real integration issue or bad tests. I realize we have flaky tests at this point but running tests before publishing/promoting is the right policy when master is used for integration. I.e. if the build fails we should not publish.

I'm concerned that by ignoring tests there is less emphasis on getting them fixed. Looking at the failure this morning after #2462 was merged:

image

We could just rerun the master build to work around this and it would have been published. We go through this rerun process for PRs with the same flaky tests. Ignoring things in master does not solve that root cause and it will still be a pain in PR branches.

Finally, if we decide to go with your changes we need to make sure that the login of publishing before seeing test results does not happen for release branches. Your patch does not handle that currently.

**Comment by travisn** [comment]
Agreed that master is the point of integration. My argument for taking this risk is that since the start of the project, I don't recall any integration tests failing because two changes went to master without being tested together. Those type of changes typically result in a merge conflict where the second PR has to rebase and rebuild. And if there was a soft conflict, we still have the build and unit tests that have to pass before we would publish.

Why do you feel tests are being ignored? I'm looking at virtually every failure in master and will continue to do so even if the publish happens first. We would all love to fix every random issue, but they are very elusive, especially the CI issues before rook code is ever run. If you don't count the CI or kubeadm issues, the integration tests themselves are actually getting quite reliable.

Rerunning the build just to publish has the side effect that more users can be impacted by having a difference between the yamls and the published master build. We've been doing this for a while and it's a headache.

Can you restate the problem you're seeing with the login before seeing the test results? I thought since the release branch had a separate promote pipeline that we were ok.

**Comment by travisn** [comment]
@bassam Could you take another look at this? The release build is delayed today because of a random failure. With this change, the release would not have been blocked. thanks

**Comment by travisn** [comment]
The Jenkinsfile has been updated such that the master build will be published as soon as the build and unit test stages are successful. The integration tests will run after the build is published. This ensures:
- The generated help documentation matches what is in master
- The latest master published build matches the yaml files in master

The longer the delay between merging and publishing, the more likely users are to run into issues with test deployments. This change minimizes the gap.

Publishing of the release branch remains unchanged, which will be published after the integrations tests pass. In the event that the tests fail and the failures are found to be known issues, the build can be triggered manually with the `skipIntegrationTests=true` flag to publish.

File: Jenkinsfile [hunk 1]
```diff
@@ -4,7 +4,7 @@ pipeline {
    parameters {
          booleanParam(defaultValue: true, description: 'Execute pipeline?', name: 'shouldBuild')
          booleanParam(defaultValue: false, description: 'Run Only Smoke Test', name: 'smokeOnly')
-
+         booleanParam(defaultValue: false, description: 'Skip Integration Tests?', name: 'skipIntegrationTests')
```
Comment (jbw976) [inline Jenkinsfile]: this new `skipIntegrationTests` parameter will show up in the Jenkins UI when you manually trigger a build, right? Are there any other ways to set it?

File: Jenkinsfile [hunk 2]
```diff
@@ -57,10 +57,29 @@ pipeline {
...
+            environment {
+                DOCKER = credentials('rook-docker-hub')
+                AWS = credentials('rook-jenkins-aws')
+                GIT = credentials('rook-github')
+            }
+            steps {
+                sh 'docker login -u="${DOCKER_USR}" -p="${DOCKER_PSW}"'
+                sh 'build/run make -j\$(nproc) -C build/release build BRANCH_NAME=${BRANCH_NAME} GIT_API_TOKEN=${GIT_PSW}'
+                sh 'build/run make -j\$(nproc) -C build/release publish BRANCH_NAME=${BRANCH_NAME} AWS_ACCESS_KEY_ID=${AWS_USR} AWS_SECRET_ACCESS_KEY=${AWS_PSW} GIT_API_TOKEN=${GIT_PSW}'
+                sh 'build/run make -j\$(nproc) -C build/release promote BRANCH_NAME=master CHANNEL=master AWS_ACCESS_KEY_ID=${AWS_USR} AWS_SECRET_ACCESS_KEY=${AWS_PSW}'
```
Comment (jbw976) [inline Jenkinsfile]: maybe add a comment here for why this auto promote is done for master.  quickly looking, it could appear that this logic is entirely duplicated between this new `Publish if Master` stage and the `Publish if Release` stage, but `promote` call is a key difference between them.  It was more obvious when the `promote` call was behind an inline check for master branch, but now since it's been moved to a different stage, that context is lost.

File: Jenkinsfile [hunk 3]
```diff
@@ -57,10 +57,29 @@ pipeline {
...
+            }
+        }
+
+        stage('Integration Tests') {
             when {
                 expression {
-                    return env.shouldBuild != "false" && env.shouldTest != "false"
+                    return env.shouldBuild != "false" && env.shouldTest != "false" && !env.skipIntegrationTests
```
Comment (jbw976) [inline Jenkinsfile]: the other checks are against string values (e.g., `"false"`), but this new one uses a boolean operator.  is there a reason for that? not an expert on Jenkinsfile syntax.

File: Jenkinsfile [hunk 4]
```diff
@@ -91,10 +110,11 @@ pipeline {
                 }
             }
         }
-        stage('Publish') {
+
+        stage('Publish if Release') {
             when {
                 expression {
-                    return env.shouldBuild != "false" && env.shouldTest != "false"
```
Comment (jbw976) [inline Jenkinsfile]: should we keep these old checks still?  why do they no longer matter?

File: Jenkinsfile (hunk 1 above)
Comment (travisn) [reply Jenkinsfile]: There aren't other ways to set it. Since we only want to use it in the case of master and release builds we don't have a PR in which to add keywords.

File: Jenkinsfile (hunk 2 above)
Comment (travisn) [reply Jenkinsfile]: yeah, will add a comment to keep that context

File: Jenkinsfile (hunk 3 above)
Comment (travisn) [reply Jenkinsfile]: Actually I'm still trying to get this condition to work. I've tried several variations including string comparisons or the boolean and they are always evaluating to false and skipping the stage. I'm a novice at the jenkins syntax as well!

File: Jenkinsfile (hunk 4 above)
Comment (travisn) [reply Jenkinsfile]: `shouldBuild` and `shouldTest` only apply to PR builds since they come from comments in the PR so they are NA for the release builds.

File: Jenkinsfile [hunk 5]
```diff
@@ -2,9 +2,7 @@
 
 pipeline {
    parameters {
-         booleanParam(defaultValue: true, description: 'Execute pipeline?', name: 'shouldBuild')
```
Comment (travisn) [inline Jenkinsfile]: These two lines were not working. Their purpose was to be used in the PR description, which is still handled below.

File: Jenkinsfile [hunk 6]
```diff
@@ -57,10 +55,30 @@ pipeline {
...
+                // automatically promote the master builds
+                sh 'build/run make -j\$(nproc) -C build/release promote BRANCH_NAME=master CHANNEL=master AWS_ACCESS_KEY_ID=${AWS_USR} AWS_SECRET_ACCESS_KEY=${AWS_PSW}'
+            }
+        }
+
+        stage('Integration Tests') {
             when {
                 expression {
-                    return env.shouldBuild != "false" && env.shouldTest != "false"
+                    return env.shouldBuild != "false" && env.shouldTest != "false" && !params.skipIntegrationTests
```
Comment (travisn) [inline Jenkinsfile]: @jbw976 This is working now. `params` is the magic jenkins keyword.
